data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/S3SinkService.java

```diff
@@ -0,0 +1,159 @@
+        numberOfRecordsFailedCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_FLUSHED_TO_S3_FAILED);
+        s3ObjectSizeSummary = pluginMetrics.summary(S3_OBJECTS_SIZE);
+    }
+
+    /**
+     * @param records received records and add into buffer.
+     */
+    void output(Collection<Record<Event>> records) {
+        reentrantLock.lock();
+        final String bucket = s3SinkConfig.getBucketOptions().getBucketName();
+        final int maxRetries = s3SinkConfig.getMaxUploadRetries();
+        if (currentBuffer == null) {
+            currentBuffer = bufferFactory.getBuffer();
+        }
+        try {
+            for (Record<Event> record : records) {
+
+                final Event event = record.getData();
+                final String encodedEvent;
+                encodedEvent = codec.parse(event);
+                final byte[] encodedBytes = encodedEvent.getBytes();
+                if (willExceedThreshold()) {
+                    s3ObjectSizeSummary.record(currentBuffer.getEventCount());
+                    LOG.info("Event collection Object info : Byte_capacity = {} Bytes," +
+                            " Event_count = {} Records & Event_collection_duration = {} Sec",
+                            maxBytes.getBytes(), currentBuffer.getEventCount(), currentBuffer.getDuration());
+                    boolean isUploadedToS3 = currentBuffer.flushToS3(s3Client, bucket, generateKey(), maxRetries);
+                    if (isUploadedToS3) {
+                        LOG.info("Event collection Object uploaded successfully");
+                        numberOfRecordsSuccessCounter.increment(currentBuffer.getEventCount());
+                        snapshotSuccessCounter.increment();
+                    } else {
+                        LOG.info("Event collection Object upload failed");
+                        numberOfRecordsFailedCounter.increment(currentBuffer.getEventCount());
+                        snapshotFailedCounter.increment();
+                    }
+                    currentBuffer = bufferFactory.getBuffer();
+                }
+                currentBuffer.writeEvent(encodedBytes);
+            }
+        } catch (NullPointerException | IOException | InterruptedException e) {
+            LOG.error("Exception while write event into buffer :", e);
+            Thread.currentThread().interrupt();
+        }
+        reentrantLock.unlock();
+    }
+
+    /**
+     * Generate the s3 object path prefix & object file name.
+     * @return object key path.
+     */
+    protected String generateKey() {
+        final String pathPrefix = ObjectKey.buildingPathPrefix(s3SinkConfig);
+        final String namePattern = ObjectKey.objectFileName(s3SinkConfig);
+        return (!pathPrefix.isEmpty()) ? pathPrefix + namePattern : namePattern;
+    }
+
+    /**
+     * Check threshold limits.
+     * @return boolean value whether the threshold are met.
+     */
+    private boolean willExceedThreshold() {
```

Comment:
This is very important logic and is currently not thoroughly tested.

This could exist in a different class. Then it would highly testable. Perhaps this can come as a follow-on PR.

Reply:
Create new class `ThresholdCheck` along with test-class and tested thoroughly.